<a href="https://colab.research.google.com/github/Abdullah2020/Machine-Learning-ZoomCamp2023/blob/master/Homework_ML_ZoomCamp1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **ML Zoomcamp 2023 - Homework #1**


In [ ]:
#!pip3 install pandas==2.1.0

In [2]:
#Import libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

print('Setup Completed...!!!')

Setup Completed...!!!


# **Question 1**
What's the version of Pandas that you installed?

In [3]:
pd.__version__

'2.1.0'

**Getting the data**

In [4]:
# URL of the dataset
url = "https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv"

# Read the data into a DataFrame
df = pd.read_csv(url)

In [5]:
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

# **Question 2**
How many columns are in the dataset?

In [6]:
print("Answer: There are {} columns in the dataset.".format(df.shape[1]))


Answer: There are 10 columns in the dataset.


## **Question 3**
Which columns in the dataset have missing values?

In [7]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

**Answer**: `total_bedrooms` contains 207 missing values.

# **Question 4**
How many unique values does the `ocean_proximity` column have?

In [8]:
df['ocean_proximity'].nunique()

5

# **Question 5**
What's the average value of the median_house_value for the houses located near the bay?

In [9]:
df.groupby('ocean_proximity').median_house_value.mean()

ocean_proximity
<1H OCEAN     240084.285464
INLAND        124805.392001
ISLAND        380440.000000
NEAR BAY      259212.311790
NEAR OCEAN    249433.977427
Name: median_house_value, dtype: float64

In [10]:
#to be more specific
near_bay = df[df['ocean_proximity'] == 'NEAR BAY']['median_house_value'].mean()
print('Answer: {}'.format(near_bay.round(1)))

Answer: 259212.3


**Answer**: Near the Bay is 259212.3

# **Question 6**
1.  Calculate the average of `total_bedrooms` column in the dataset.
2. Use the fillna method to fill the missing values in `total_bedrooms` with the mean value from the previous step.
3. Now, calculate the average of `total_bedrooms` again.
4. Has it changed?

Has it changed?

In [11]:
#1. The initial average of total_bedrooms

df_mean1 = df['total_bedrooms'].mean()
print("The average values of total_bedrooms is: ", df_mean1.round(3))

The average values of total_bedrooms is:  537.871


In [12]:
#2- using fillna method

df['total_bedrooms'] = df['total_bedrooms'].fillna(value=df_mean1)

In [13]:
#3-
df_mean2 = df['total_bedrooms'].mean()
print("The 2nd average values of total_bedrooms is: ", df_mean2.round(3))

The 2nd average values of total_bedrooms is:  537.871


**Answer**: #4- There is no change in the average after filling the missing values using fillna method.

# **Question 7**
1. Select all the options located on islands.
2. Select only columns `housing_median_age`, `total_rooms`, `total_bedrooms`.
3. Get the underlying NumPy array. Let's call it `X`.
4.  Compute matrix-matrix multiplication between the transpose of `X` and X`X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
5. Compute the inverse of `XTX`.
6. Create an array `y` with values `[950, 1300, 800, 1000, 1300]`.
7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
8. What's the value of the last element of `w`?

In [14]:
#1 and 2.
selected_columns = df[df['ocean_proximity']=='ISLAND'][['housing_median_age', 'total_rooms', 'total_bedrooms']]


In [15]:
# 3. Get the underlying NumPy array. Let's call it X.
X = selected_columns.to_numpy()
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [16]:
# 4.Compute XTX (matrix-matrix multiplication)
XTX = np.dot(X.T, X)
XTX


array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [17]:
#5. Compute the inverse of XTX
XTX_inverse = np.linalg.inv(XTX)
XTX_inverse


array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [18]:
#6. Create array y
y = np.array([950, 1300, 800, 1000, 1300])
y

array([ 950, 1300,  800, 1000, 1300])

In [19]:
#7. Compute w
w = np.dot(np.dot(XTX_inverse, X.T), y)

# Get the last element of w
last_w_index = w[-1]

print(last_w_index)

5.699229455065586
